File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] h) {
        
        
        Stack<Integer> st = new Stack<>(); 
        int rain = 0 ; 
        for(int r =0 ; r < h.length ; r++){

            while( !st.isEmpty() && h[st.peek()] < h[r] ){ 
                int btm = st.pop();
                if( st.isEmpty() ) {
                    break ; } 

                int l = st.peek() ; 
                int wid = (r-l -1) ; 

                rain += wid*( Math.min( h[r] , h[l] ) - h[btm]) ;
            }
            st.push(r) ; 
        }
        return rain ; 
    }
}

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& h) {
        
         
        int rain = 0 ; 
        int l =1 , r = h.size()-1 ; 
        int maxl = h[0] ,  maxr = h[h.size()-1] ; 
        while( l <=r ) { 

            if( maxl < maxr){ 
                if( h[l] > maxl) maxl = h[l] ; 
                else {
                    rain += maxl-h[l] ; } 
                l++ ; 
            }
            else { 
                if( h[r] > maxr) maxr = h[r] ; 
                else { 
                    rain += maxr - h[r] ; } 
                r -- ;
            }
        }
        return rain ;  
    
        
    }
};